Coinduction handling for recursive solver (#698)
rust-lang/chalk: book/src/recursive/coinduction.md

# Coinduction

This sub-chapter is meant to describe the current handling of coinductive goals in the recursive solver rather than providing an extensive overview over the theoretical backgrounds and ideas.
It follows the description in [this GitHub comment](https://github.com/rust-lang/chalk/issues/399#issuecomment-643420016) and the Zulip topic linked there.

## General Idea
The general idea for the handling of coinductive cycles in the recursive solver is to start by assuming the goal is provable and then try to find evidence that it is not.
This search for a disproof is done by the standard recursive solving process described in the sub-chapters before.

> Albeit this approach would allow for the handling of mixed inductive/co-inductive cycles, these are actually handled as errors to prevent propagation of the assumed provability outside of the coinductive cycle.

Reviewer: See my explanation below. Maybe we can add a subsection like

## Mixed co-inductive and inductive cycles

and then reference that section from the code? My comment seems to long to include in the code itself.

Author: I tried adding a comprehensive subsection about mixed cycles. I hope that it is understandable.

This propagation of the assumed solution might also happen in pure coinductive cycles and can potentially lead to invalid results.

## Prevention of Invalid Results

> The problem of invalid results propagated outside of the coinductive cycle is also described in the [Coinduction chapter](../engine/logic/coinduction.md) for the SLG solver alongside the rather complex handling used with it.

Reviewer: It's worth noting that, in the recursive solver, no special cases are required. The logic for handling inductive cycles suffices.

### The Problem
The problem arises if a solution that is purely based on the positive starting value for the coinductive cycle is cached and as such propagated to other goals that are possibly reliant on this. An example may look like this (cf. the test case `coinduction::coinductive_unsound`):

```notrust
C :- C1.
C :- C2
C1 :- C2, C3.
C2 :- C1.
```

> Here `C` may be proved by either showing `C1` or `C2`.

Reviewer: I suspect this would be more readable if formatted as a bulleted list

Assuming the solver starts evaluating the branch with `C1` first, it then recursively tries to prove `C2` and `C3`.
For proving `C2` it needs to show `C1` again, the coinductive cycle becomes evident.
Therefore, `C1` is assumed to be provable and the solver proves `C2` with this information.
Assuming, the solve does not handle this case specifically, the solution for `C2` is cached. 
Now it tries solving `C3` but fails due to the lack of information about it.
As such, `C1` can also not be proven for this program.
The recursive solver will now attempt to prove the initial goal `C` by solving `C2`.
Unfortunately, it finds the invalidly cached solution and returns it as proof for `C`.

> By visualizing this path of computation, it becomes evident, where the problem lies:

Reviewer: I guess we sort of did that here... maybe the paragraph above and this can be combined? There's a bit of extra info in the paragraph above (e.g., "assuming the solver does not handle this case specifically, the solution for `C2` is cached" is more informative than "Store this result about `C2` in the cache")

The only difference is the start value for coinductive goals.
Whereas inductive goals start with a negative result and are iterated until a least fixed-point is found, coinductive goals start with a positive result (i.e. a unique solution with identity substitution).
This negative result is then iterated until a greatest fixed-point is reached.

## Mixed co-inductive and inductive Cycles
As described above, the handling of inductive and coindutive cycles differs only in the start value from which the computation begins.
Thus, it might seem reasonable to have mixed inductive and coinductive cycles as all goals inside these cycles would be handled the same way anyway.
Unfortunately, this is not possible for the kind of logic that Chalk is based on (i.e. essentially an extension of co-LP for Hereditary Harrop clauses, cf. [this paper][co-LP]).

There is fundamental difference between results for inductive cycles and results for coinductive cycles of goals.
An inductive goal is provable if and only if there exists a proof for it consisting of a finite chain of derivations from axioms that are members of the least-fixed point of the underlying logic program.

> On the other hand, coinductive goals are provable if and only if there exists an infinite derivation starting from the axioms that proves it.

Author: I adjusted this and the other suggested changes. 

More of a meta question: am I supposed to resolve the review comments that were fixed already or is this something the reviewer does? It's currently a bit unclear to me which of these are still relevant.

> On the other hand, coinductive goals are provable if and only if there exists an infinite derivation starting from the axioms that proves it.

Reviewer: I usually resolve if it's a nit but otherwise leave it up to the reviewer.

This difference between inductive and coinductive results becomes a problem when combined in a single cycle.
Consider a coinductive goal `CG` and an inductive goal `IG`. Now consider the simplest possible mixed cycle:
```notrust
CG :- IG
IG :- CG
```
It is apparent, that there can not exist a solution for `IG` as the cyclic dependency prevents a finite proof derivation.
In contrast to that, `CG` could potentially be provable as the derivation *`CG` if `IG` if `CG` if `IG` ...* is infinite and based only on the two axioms.
As a result, `CG` would hold whereas `IG` would not hold, creating a contradiction.

The simplest solution to this problem, proposed by Simon et al. in [their paper about co-LP][co-LP], is to disallow mixed inductive and coinductive cycles.
This approach is also used by Chalk.

## Prevention of Invalid Results
The problem of invalid results propagated outside of the coinductive cycle is also described in the [Coinduction chapter](../engine/logic/coinduction.md) for the SLG solver alongside the rather complex handling used with it.

> Whereas the SLG solver introduces a new kind of clauses (so called ExClauses) to handle coinduction, it is sufficient for the recursive solver to use the same logic for inductive and coinductive cycles.

Reviewer: ```suggestion
Whereas the SLG solver introduces [special constructs](../engine/logic/coinduction.html#nikos-proposed-solution) to handle coinduction, it is sufficient for the recursive solver to use the same logic for inductive and coinductive cycles.
```
